libr/bin/format/elf/elf.c

```diff
@@ -405,39 +408,53 @@ static void store_versioninfo_gnu_versym(struct Elf_(r_bin_elf_obj_t) *bin, Elf_
 	free (data);
 }
 
-static void store_versioninfo_gnu_verdef(struct Elf_(r_bin_elf_obj_t) *bin, Elf_(Shdr) *shdr, int sz) {
-	const char *section_name = &bin->shstrtab[shdr->sh_name];
+static void store_versioninfo_gnu_verdef(struct Elf_(r_bin_elf_obj_t) *bin, Elf_(Shdr) *shdr) {
+	const char *section_name = NULL; 
+	if (shdr->sh_name > bin->shstrtab_size)
+		return;
+	section_name = &bin->shstrtab[shdr->sh_name];
 	eprintf ("Version definition section '%s' contains %d entries:\n", section_name, shdr->sh_info);
 }
 
-static void store_versioninfo_gnu_verneed(struct Elf_(r_bin_elf_obj_t) *bin, Elf_(Shdr) *shdr, int sz) {
-	ut8 *need = malloc (shdr->sh_size);
-	const char *section_name = &bin->shstrtab[shdr->sh_name];
+static void store_versioninfo_gnu_verneed(struct Elf_(r_bin_elf_obj_t) *bin, Elf_(Shdr) *shdr) {
+	int sz = shdr->sh_size;
+	ut8 *need = NULL;
+	ut8 *vend = NULL;
+	const char *section_name = NULL;
 	int i;
 	int cnt;
+	if (shdr->sh_name > bin->ehdr.e_shnum)
+		return;
+	need = malloc (sz);
+	if (!need) return;
+	section_name = &bin->shstrtab[shdr->sh_name];
 	eprintf ("Version needs section '%s' contains %d entries:\n", section_name, shdr->sh_info);
 	eprintf (" Addr: 0x%08"PFMT64x, (ut64)shdr->sh_addr);
 	eprintf (" Offset: 0x%08"PFMT64x"  Link to section: %x (%s)\n",
 		(ut64)shdr->sh_offset, shdr->sh_link, section_name);
-	//int num_verneed = shdr->sh_size / sizeof (Elf_(Verneed));
-	if (r_buf_read_at (bin->b, shdr->sh_offset, need, shdr->sh_size) != shdr->sh_size) {
+	if (shdr->sh_offset > bin->size || shdr->sh_offset + sz > bin->size)
+		return;
+	if (shdr->sh_offset + sz < sz)
+		return;
+	if (r_buf_read_at (bin->b, shdr->sh_offset, need, sz) != sz) {
 		eprintf ("Cannot read section headers\n");
 		free (need);
 		return;
 	}
-	for (i = 0, cnt = 0; i<sz && cnt < shdr->sh_info; ++cnt) {
+	vend = need + sz;
+	//XXX we should use DT_VERNEEDNUM instead of sh_info
+	//TODO https://sourceware.org/ml/binutils/2014-11/msg00353.html
```

Comment:
@0x2Adr1 if you are going to improve the code of version info instead of sh_info use DT_VERNEEDNUM, because it seems there is discrepancy between them  and it can lead to security issues and bad parsing